diff -rup a/lib/tags/demo.js b/lib/tags/demo.js
--- a/lib/tags/demo.js	2015-08-05 22:43:25.196178530 +0530
+++ b/lib/tags/demo.js	2015-08-05 22:57:42.916754182 +0530
@@ -33,7 +33,7 @@
 	module.exports = {
 		add: function(  line, curData, scope, objects, currentWrite ) {
 			
-			var m = line.match(/^\s*@demo\s*([\w\.\/\-\$]*)\s*([\w]*)/)
+			var m = line.match(/^\s*@demo\s*([\w\.\/\-\$#]*)\s*([\w]*)/);
 			if ( m ) {
 				var src = m[1] ? m[1].toLowerCase() : '';
 				var heightAttr = m[2].length > 0 ? " data-demo-height='" + m[2] + "'" : '';
