diff --git a/app/Plugin/FileStorage/Event/ImageProcessingListener.php b/app/Plugin/FileStorage/Event/ImageProcessingListener.php
index bbd0ddacd..9ab50a7fc 100644
--- a/app/Plugin/FileStorage/Event/ImageProcessingListener.php
+++ b/app/Plugin/FileStorage/Event/ImageProcessingListener.php
@@ -363,7 +363,7 @@ class ImageProcessingListener extends Object implements CakeEventListener {
 				mkdir(TMP . 'image-processing');
 			}
 
-			$tmpFile = TMP . 'image-processing' . DS . String::uuid();
+			$tmpFile = TMP . 'image-processing' . DS . CakeText::uuid();
 			file_put_contents($tmpFile, $Storage->read($path));
 
 			return $tmpFile;
diff --git a/app/Plugin/FileStorage/Event/LocalImageProcessingListener.php b/app/Plugin/FileStorage/Event/LocalImageProcessingListener.php
index 39fbf645f..80e159f11 100644
--- a/app/Plugin/FileStorage/Event/LocalImageProcessingListener.php
+++ b/app/Plugin/FileStorage/Event/LocalImageProcessingListener.php
@@ -197,7 +197,7 @@ class LocalImageProcessingListener extends Object implements CakeEventListener {
 				mkdir(TMP . 'image-processing');
 			}
 
-			$tmpFile = TMP . 'image-processing' . DS . String::uuid();
+			$tmpFile = TMP . 'image-processing' . DS . CakeText::uuid();
 			$imageData = $Storage->read($path);
 
 			file_put_contents($tmpFile, $imageData);
diff --git a/app/Plugin/FileStorage/Model/FileStorage.php b/app/Plugin/FileStorage/Model/FileStorage.php
index 992d485b0..a36d23768 100644
--- a/app/Plugin/FileStorage/Model/FileStorage.php
+++ b/app/Plugin/FileStorage/Model/FileStorage.php
@@ -250,7 +250,7 @@ class FileStorage extends FileStorageAppModel {
 			new Folder($path, true);
 		}
 
-		return $path . String::uuid();
+		return $path . CakeText::uuid();
 	}
 
 /**
