Index: passbolt-@PASSBOLT_FLAVOUR@-server/webroot/index.php
===================================================================
--- passbolt-@PASSBOLT_FLAVOUR@-server.orig/webroot/index.php
+++ passbolt-@PASSBOLT_FLAVOUR@-server/webroot/index.php
@@ -16,7 +16,7 @@
  */
 
 // Check platform requirements
-require dirname(__DIR__) . '/config/requirements.php';
+include '/etc/passbolt/requirements.php';
 
 // For built-in server
 if (PHP_SAPI === 'cli-server') {
